src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,149 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T item) {
+            this.item = item;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (isEmpty()) {
+            head = tail = newNode;
+        } else {
+            insertAtTheEndOfTheList(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Objects.checkIndex(index, size + 1);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            insertAtTheBeginningOfTheList(value);
+        } else {
+            insertAtTheMiddleOfTheList(value, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> getElement = findingNodeByIndex(index);
+        return getElement.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> overwriteValue = findingNodeByIndex(index);
+        T oldValue = overwriteValue.item;
+        overwriteValue.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Objects.checkIndex(index, size);
+        Node<T> removedNode = findingNodeByIndex(index);
+        return unlink(removedNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.item == object || (object != null && object.equals(currentNode.item))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private Node<T> findingNodeByIndex(int index) {
```

Comment:
[checklist](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa): Don’t iterate your nodes from the head to tail or vice versa.